product_category_company/views/product_template_view.xml

```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<!-- Copyright (C) 2022 Akretion (<http://www.akretion.com>).
+     @author Kévin Roche <kevin.roche@akretion.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record id="product_template_view_form" model="ir.ui.view">
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_form_view" />
+        <field name="arch" type="xml">
+            <field name="categ_id" position="attributes">
+                <attribute
+                    name="domain"
+                >['|', ('company_id', '=', False), ('company_id', '=', company_id)]</attribute>
+
```

Comment:
Workaround

`    company_id = fields.Many2one(
        comodel_name="res.company", default=lambda s: s.env.user.company_id.id
    )
`

Reply:
> when I create a product, there is no company_id in default/creating data, then company_id = False

This seems like an Odoo bug in a multi-company database to me. But I'm like 80% sure that I'm misunderstanding this, because I doubt that there would be a bug _that_ obvious in Odoo.

Your proposed workaround would work for new product records. Old records would still need to be corrected. (Although `env.user.company_id` has been changed to `env.company` since Odoo 13.)

I'll write a test first, and see what actually happens.

Reply:
Thanks for your explanation. Maybe you're right, It's probably a bug in odoo. In this case it shouldn't be resolved in this PR !?

Reply:
@legalsylvain do you an idea here as multi company intensive user ?

Reply:
Wrote a test in [0d021b4](https://github.com/OCA/multi-company/pull/470/commits/0d021b4b83160d64485ff5b74388551f9c388655). Your analysis was right.

I'm still doubtful that this is actually a bug, though. Surely it's intended behaviour. Maybe products are supposed to be shared between companies? And if so, then this module is not the correct place to limit that behaviour.

Reply:
Thanks a lot @carmenbianca , nice job

Reply:
Maybe you can only set this default value only for not multicompany user to avoid to break some tests ?

Reply:
I'll await @legalsylvain 's sage input, unless there's a rush.

Reply:
> Hi @carmenbianca , I have an other requirement here, sorry. It's an improvement, not a migration topic but I think it can be useful for everybody. Maybe a test could be added:
> 
> * I'm a user without "base.group_multi_company" group (mono company user)
> * when I create a product, there is no company_id in default/creating data, then company_id = False
> * domain in categ_id doesn't allow me to select category with a company_id (matching with self.env.user.company_id) => it Should allow to select a category of my company.
> 
> Do you know how could we fix that ?
> 
> Please @carmenbianca Do you confirm this case
> 
> Thanks a lot.

@bealdav. sorry for my late comment. it's just an additional comment.  for general purpose,  in V16 already enought as mention by @legalsylvain. I mean for your case, in real business flow in a company, once the rules of system in internal management of a company use multi company, all users must be set as multi company, no user in single company.  Just my opinion.